### PR TITLE
Added CQL paddle version

### PR DESCRIPTION
**Pull request by Jiukaishi**

### CODE REVIEW

**Comment by zenghsh3** [comment]
@Jiukaishi You can fix code style issues by following the guide: https://github.com/PaddlePaddle/PARL/blob/96070194022fb044dd0be2c9e0a5ffc24b47212c/docs/zh_CN/Overview.md#%E4%BB%A3%E7%A0%81%E8%B4%A1%E7%8C%AE%E8%A7%84%E8%8C%83

File: examples/CQL/train.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    rpm.load_from_d4rl(d4rl.qlearning_dataset(env))
+
+    total_steps = 0
+    test_flag = 0
+    for total_steps in trange(int(args.train_total_steps)):
+
+        # Train steps
+        batch_obs, batch_action, batch_reward, batch_next_obs, batch_terminal = rpm.sample_batch(
+            BATCH_SIZE)
+        closs, aloss = agent.learn(batch_obs, batch_action, batch_reward,
```
Comment (zenghsh3) [inline examples/CQL/train.py]: recommended naming: c_loss/critic_loss and a_loss/actor_loss

File: parl/algorithms/paddle/cql.py [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        assert isinstance(lagrange_thresh, float)
+        assert isinstance(min_q_version, int)
+        assert isinstance(min_q_weight, float)
+        assert isinstance(alpha, float)
+
+        self.gamma = gamma
+        self.tau = tau
+        self.actor_lr = actor_lr
+        self.critic_lr = critic_lr
+        self.sample_normal = Normal(loc=0., scale=1.)
```
Comment (zenghsh3) [inline parl/algorithms/paddle/cql.py]: It's unused, pls remove it.

File: parl/algorithms/paddle/cql.py [hunk 3]
```diff
@@ -0,0 +1,299 @@
...
+        self.with_lagrange = with_lagrange
+        self.lagrange_thresh = lagrange_thresh
+        self.min_q_version = -1 if self.with_lagrange else min_q_version
+        self.min_q_weight = min_q_weight
+        self.alpha = alpha
+        self.temp = 1.0
+        self.num_random = 10
+        self._current_steps = 0
+
+        #self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
```
Comment (zenghsh3) [inline parl/algorithms/paddle/cql.py]: Pls remove these comments.

File: parl/algorithms/paddle/cql.py [hunk 4]
```diff
@@ -0,0 +1,299 @@
...
+
+    def sample(self, obs):
+        """ Define the sampling process. This function returns an action with noise to perform exploration.
+        """
+        act_mean, act_log_std = self.model.policy(obs)
+        normal = Normal(act_mean, act_log_std.exp())
+        # for reparameterization trick  (mean + std * N(0, 1))
+        x_t = normal.sample(
+            [1]
+        )  #act_mean + self.sample_normal.sample(act_log_std.shape) * act_log_std.exp()
```
Comment (zenghsh3) [inline parl/algorithms/paddle/cql.py]: Pls remove these comments.

File: parl/algorithms/paddle/cql.py [hunk 5]
```diff
@@ -0,0 +1,299 @@
...
+                                      q2_next)  # a little different from SAC
+            target_Q = reward + self.gamma * (1. - terminal) * target_Q
+        cur_q1, cur_q2 = self.model.value(obs, action)
+        qf1_loss = F.mse_loss(cur_q1, target_Q)
+        qf2_loss = F.mse_loss(cur_q2, target_Q)
+
+        ## add CQL
+        random_actions_tensor = paddle.uniform(
+            shape=[cur_q2.shape[0] * self.num_random, action.shape[-1]])
+        # torch.FloatTensor(
```
Comment (zenghsh3) [inline parl/algorithms/paddle/cql.py]: Pls remove these comments.

**Review by zenghsh3 — CHANGES_REQUESTED** [review]

File: examples/CQL/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+## Reproduce CQL with PARL
+
+Different from Online RL which gather new data directly from the environment, Offline RL algorithms learn a policy or Q-function from static dataset, without additional interaction with the environment.
```
Comment (TomorrowIsAnOtherDay) [inline examples/CQL/README.md]: Remove the sentence about the offline RL. We would like the reader to refer to the original paper for more introduction.

File: examples/CQL/README.md [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+### Benchmark result
+![avatar](./result.png)
+
+
+
+
+## How to use
+### Dependencies:
++ python3.5+
++ [parl>2.0.2](https://github.com/PaddlePaddle/PARL)
```
Comment (TomorrowIsAnOtherDay) [inline examples/CQL/README.md:18]: \>=2.0.4

File: examples/CQL/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
+## Reproduce CQL with PARL
+
+Different from Online RL which gather new data directly from the environment, Offline RL algorithms learn a policy or Q-function from static dataset, without additional interaction with the environment.
+
+Based on PARL, the CQL algorithm of deep reinforcement learning has been reproduced, reaching the same level of indicators as the paper on continuous control datasets from the D4RL benchmark.
+
+> Paper: CQL in [Conservative Q-Learning for Offline Reinforcement Learning](https://arxiv.org/abs/2006.04779)
+
+### Env and dataset introduction
++ D4RL datasets: Please see [here](https://sites.google.com/view/d4rl/home) to know more about D4RL datasets. D4RL require Mujoco as a dependency. For more D4RL usage methods, please refer to its [guide](https://github.com/rail-berkeley/d4rl#using-d4rl).
```
Comment (TomorrowIsAnOtherDay) [inline examples/CQL/README.md]: The algorithm is tested in the [D4RL dataset](https://sites.google.com/view/d4rl/home), one of the most commonly used dataset for offline RL. For more details about D4RL usagage, please refer to [the page](https://github.com/rail-berkeley/d4rl#using-d4rl).

File: examples/CQL/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
++ Mujoco simulator: Please see [here](http://mujoco.org/) to know more about Mujoco simulator and obtain a license.
+
+### Benchmark result
+![avatar](./result.png)
```
Comment (TomorrowIsAnOtherDay) [inline examples/CQL/README.md]: @zenghsh3 Please tell @Jiukaishi  how to upload the result image to the other repository.

File: parl/algorithms/paddle/cql.py [hunk 10]
```diff
@@ -0,0 +1,295 @@
...
+            policy_eval_start (int): try doing behaivoral cloning at the beginning, 40000 or 10000 work similarly
+            with_automatic_entropy_tuning (bool): train with automatic entropy tuning in Actor.
+            with_lagrange (bool): train with lagrange
+            lagrange_thresh (float): the value of tau, corresponds to the CQL(lagrange) version, suggest 10.0 in mujoco and 5.0 in Franka kitchen or Adroit domains
+            min_q_version (int): min_q_version = 3 (CQL(H)), = 2 (CQL(rho)), will be set to <0 in cql if not using lagrange
+            min_q_weight (float): the value of alpha in Critic loss, suggest 5.0 or 10.0 if not using lagrange
+            alpha (float): the value of alpha(temperature parameter) in Actor loss, determines the relative importance of entropy term against the reward
+        """
+        # checks
+        # check_model_method(model, 'value', self.__class__.__name__)
```
Comment (TomorrowIsAnOtherDay) [inline parl/algorithms/paddle/cql.py]: Why do we comment these lines? 
I think they are necessary check to reuse the algorithm?

**Review by TomorrowIsAnOtherDay — APPROVED** [review]
LGTM.

**Review by zenghsh3 — APPROVED** [review]
LGTM

**Pull request merged**